src/Vlingo.Common/BasicCompletes.cs

```diff
@@ -0,0 +1,569 @@
+        {
+            state.FailedValue(failedOutcomeValue);
+            state.Action(Action<T>.With(consumer));
+            if (state.IsCompleted)
+            {
+                state.CompleteActions();
+            }
+            else
+            {
+                state.StartTimer(timeout);
+            }
+            return this;
+        }
+
+        public virtual ICompletes<T> AndThenConsume(long timeout, System.Action<T> consumer)
+            => AndThenConsume(timeout, default(T), consumer);
+
+        public virtual ICompletes<T> AndThenConsume(T failedOutcomeValue, System.Action<T> consumer)
+            => AndThenConsume(-1L, failedOutcomeValue, consumer);
+
+        public virtual ICompletes<T> AndThenConsume(System.Action<T> consumer) => AndThenConsume(-1L, default(T), consumer);
+
+        public virtual O AndThenInto<F, O>(long timeout, F failedOutcomeValue, Func<T, O> function)
+        {
+            var nestedCompletes = new BasicCompletes<O>(state.Scheduler);
+            nestedCompletes.state.FailedValue(failedOutcomeValue);
+            nestedCompletes.state.FailureAction((BasicCompletes<O>.Action<O>)(object)state.FailureActionFunction());
+            state.Action((Action<T>)(object)BasicCompletes<O>.Action<O>.With(function, nestedCompletes));
+            if (state.IsCompleted)
+            {
+                state.CompleteActions();
+            }
+            else
+            {
+                state.StartTimer(timeout);
+            }
+            return (O)(object)nestedCompletes;
+        }
+
+        public virtual O AndThenInto<F, O>(F failedOutcomeValue, Func<T, O> function)
+            => AndThenInto(-1L, failedOutcomeValue, function);
+
+        public virtual O AndThenInto<O>(long timeout, Func<T, O> function)
+            => AndThenInto<object,O>(timeout, null, function);
+
+        public virtual O AndThenInto<O>(Func<T, O> function)
+            => AndThenInto<object, O>(-1L, null, function);
+
+        public virtual ICompletes<T> Otherwise(Func<T, T> function)
+        {
+            state.FailureAction(Action<T>.With(function));
+            return this;
+        }
+
+        public virtual ICompletes<T> OtherwiseConsume(System.Action<T> consumer)
+        {
+            state.FailureAction(Action<T>.With(consumer));
+            return this;
+        }
+
+        public virtual ICompletes<T> RecoverFrom(Func<Exception, T> function)
+        {
+            state.ExceptionAction(function);
+            return this;
+        }
+
+        public virtual T Await() => Await(-1L);
+
+        public virtual T Await(long timeout)
+        {
+            long countDown = timeout;
+            while (true)
+            {
+                if (IsCompleted)
+                {
+                    return Outcome;
+                }
+                try
+                {
+                    Thread.Sleep(TimeSpan.FromMilliseconds((countDown >= 0 && countDown < 100) ? countDown : 100));
```

Comment:
Thread.Sleep should be avoided at all costs. There are too many things it can has an impact on. Depending on the platform it is almost guaranteed to block for more than `n` milliseconds. The likelihood that your thread will re-awaken exactly after `n` milliseconds is about as impossible as impossible can be. As threads are expensive resources this makes any waiting thread a **huge** waste.

Reply:
Understood and agreed. This again goes to the discussion of porting vs. improving. I have concerns about using Thread.Sleep() as well and we will be replacing these codes using much safer alternatives. For now, consider it just a porting to c#. I have created #17 to track this later.

Reply:
Thanks @zpbappi. We'll address them later on. The goal is like you did, to track it first.